pkg/issuer/acme/dns/azuredns/azuredns.go

```diff
@@ -71,6 +73,31 @@ func NewDNSProviderCredentials(environment, clientID, clientSecret, subscription
 	}, nil
 }
 
+func getWorkloadIdentity(env azure.Environment) (*adal.ServicePrincipalToken, error) {
+
+	awiClientId := os.Getenv("AZURE_CLIENT_ID")
+	awiTenantId := os.Getenv("AZURE_TENANT_ID")
+
+	jwtBytes, err := ioutil.ReadFile(os.Getenv("AZURE_FEDERATED_TOKEN_FILE"))
+	if err != nil {
+		return nil, fmt.Errorf("Failed to get Azure Workload Identity token for file: %v", err)
+	}
```

Comment:
If the `AZURE_*` environment variables are empty or unset, these errors might be quite confusing.
If AZURE_FEDERATED_TOKEN_FILE is empty or not set, will it be clear from this error what has gone wrong?

Perhaps there needs to be some early checks that all these env vars are set and not-empty.

Can AZURE_CLIENT_ID and AZURE_TENNANT_ID be empty or unset for any reason?

Reply:
Theoretically, ClientId is not required as an annotation on the service account when using Azure Workload. This means the variable can be absent. In that case it's up to the application to pass it in manually (obtained via other means).

Keda uses this approach wherein they allow you to override the client id on a per usage basis. In context here, suppose you want to use different MSI to authenticate with different azure dns zones from the same cluster. If you place the clientId annotation on the certmanager service account, that's what the env var would be in the certmanager pod... but you'd need a way to specify the other MSI. In such a situation, the likely approach would be to not set the annotation(and thus the env var) but instead pass it in via the dns01 stanza similar to MSI now.

Perhaps those cases have been covered here.. I'm honestly not sure but figured I'd call it out based on your question j.i.c.

Reply:
Not sure that is true. The Workload Identity webhook needs the Client ID (which is set as an annotation on the service account) in order to resolve the identity and request the token. It then injects that token, as well as the env vars AZURE_FEDERATED_TOKEN_FILE, AZURE_CLIENT_ID and AZURE_TENANT_ID.

If those are not set then the Workload Identity didn't get attached correctly (or some other process overwrote them)

Reply:
> Not sure that is true. The Workload Identity webhook needs the Client ID

I can assure you that it's 100% optional. We use it in such a fashion. The only required metadata is the `azure.workload.identity/use`  _label_ on the `ServiceAccount` which is what the mutating webhook actually triggers off of. See the list of annotations/labels [here](https://learn.microsoft.com/en-us/azure/aks/workload-identity-overview#service-account-labels-and-annotations). Note that nothing is marked as required except for the label. If omitted only the tenantId/token-file environment variables are injected into the pod. Don't take my word though -- it's easy enough to verify

The `ClientId` _as a whole_ is required in the sense that by the time you make the request to AZWI you must have one. The application must look to the environment variable as a _default/fallback_ (or the only value, assuming the application offers no means by which to customize it).

The annotation is optional because you might be in a scenario where the **single** service account needs to act as **multiple** MSIs which is a supported scenario by AZWI and something they promote. This is exactly how we use it with Keda; the **single** Keda pod needs to authenticate with **multiple** resources as **different** MSIs. For example, Keda scales application `A1`  based on Storage account `S1` but scales application `A2` based on Storage account `S2`. With Pod Identity you needed to run Keda under a single "god MSI" that had access to both sets of storage. With AZWI, however, the point is you should be working in a mode of _least-privilege_ meaning independent MSIs scoped to the individual applications and their resources. If you were to annotate the KEDA `ServiceAccount` with the client ID for `A1/S1` then processing for `A2` can gain inadvertent access to unowned resources, defeating the purpose.

In cases where multiple MSI's may be bound to a single SA\* you'd more than like _not_ want a default value. Instead, you'd leave the annotation _blank_ and rely on the application to provide an alternative means by which to provide the ClientId. In Keda it's an extra field on their `TriggerAuthentication` CRD; for cert-manager it would be in `dns01.azure` stanza similar to how it is now for managed identities now. 

This is all of course assuming you _want_ to enable the scenario. IMO it's half-baked without it.

\* In Azure, federated credentials are a sub-resource of an MSI. In the context of AKS, each credential is bound to a specific cluster + service account. One Service Account can therefore have associated credentials for many MSIs. The token injected into the pod by the mutating webhook from the OIDC issuer (the cluster) is for the SA and itself MSI agnostic. That token is then exchanged with AAD, alongside an explicit client Id

Reply:
Interesting, thanks for the explanation!

Reply:
My apologies for the wall of text. We were early adopters of AZWI for AKS as part of the private preview and as you can see I'm rather passionate about it. I too thought the annotation was required before working with keda and realizing it was just a convenience.

A per-CRD override makes sense for Keda. I also think it makes sense for cert-manager where multiple identities are already supported if you provide auth settings manually. I would think you could leverage the existing clientid as an override somehow. 

Conversely, it doesn't make sense for something like external-dns. While true that app can work with multiple dns zones, those options along with identity are specified at a controller level and *not* at the CRD level as they are here.

N.B. I personally don't need this as I have one-zone-per-cluster. I also don't know how common the scenario is. But otherwise you're limiting the functionality with respect to what already exists, when the whole point is to enable  something _better_.

Reply:
@pinkfloydx33 thanks for that "wall", it was insightful :)

I've just looked through the [webhook code](https://github.com/Azure/azure-workload-identity/blob/fa921332812a4aca319ef683880a98dfd893532c/pkg/webhook/webhook.go) and saw that the webhook won't try to override env variables if they're already present.

```go
// mutateContainers mutates the containers by injecting the projected
// service account token volume and environment variables
func (m *podMutator) mutateContainers(containers []corev1.Container, clientID string, tenantID string, skipContainers map[string]struct{}) []corev1.Container {
	for i := range containers {
		// container is in the skip list
		if _, ok := skipContainers[containers[i].Name]; ok {
			continue
		}
		// add environment variables to container if not exists
		containers[i] = addEnvironmentVariables(containers[i], clientID, tenantID, m.azureAuthorityHost)
		// add the volume mount if not exists
		containers[i] = addProjectedTokenVolumeMount(containers[i])
	}
	return containers
}

// addEnvironmentVariables adds the clientID, tenantID and token file path environment variables needed for SDK
func addEnvironmentVariables(container corev1.Container, clientID, tenantID, azureAuthorityHost string) corev1.Container {
	m := make(map[string]string)
	for _, env := range container.Env {
		m[env.Name] = env.Value
	}
	// add the clientID env var
	if _, ok := m[AzureClientIDEnvVar]; !ok {
		container.Env = append(container.Env, corev1.EnvVar{Name: AzureClientIDEnvVar, Value: clientID})
	}
[...]
```

And since `Annotations` is `map[string]string`, the webhook safely assumes the value would be empty in the worst case:
```go
// getClientID returns the clientID to be configured
func getClientID(sa *corev1.ServiceAccount) string {
	return sa.Annotations[ClientIDAnnotation]
}
```

=>
```
# env | grep AZURE
AZURE_AUTHORITY_HOST=https://login.microsoftonline.com/
AZURE_CLIENT_ID=
AZURE_FEDERATED_TOKEN_FILE=/var/run/secrets/azure/tokens/azure-identity-token
AZURE_TENANT_ID=ab1cde2f-g345-6789-12kl-mno3456pr789
```

Reply:
The first line in your grep result `AZURE_AUTHORITY_HOST` is also a feature of AZWI. For public cloud it's always going to be `https://login.microsoftonline.com`, but for government, foreign clouds (and others?) it will be different. 

That value is also _supposed_ to be used when requesting a token via AZWI. See it's mention mid-way through [this article](https://learn.microsoft.com/en-us/samples/azure-samples/azure-ad-workload-identity/azure-ad-workload-identity/) and in [Keda's AZWI implementation](https://github.com/kedacore/keda/blob/77f31b747ff5dd054fc09c6ba9c53f33ff397448/pkg/scalers/azure/azure_aad_workload_identity.go#L41-L53). I can't explain why it might not be present in the sidecar code nor why Keda explicitly needs it. Perhaps it's needed in the latter due to differences in ADAL vs MSAL clients? I don't know if it needs to be investigated.

The point behind my notes above still stands: the PR as-is doesn't allow the operator to use AZWI and specify an MSI/clientId on a per-CRD/solver basis. You're locked into a single identity and thus the feature is deficient when compared to existing pod-identity and "manual" MSI/SP support.  For parity, an *optional* client Id should be passed through to `getWorkloadIdentity` which should fallback to the environment variable--`azureDNS.managedIdentity.clientID` seems to be a good choice

Reply:
@pinkfloydx33 do you think about something like that: https://github.com/cert-manager/cert-manager/compare/master...weisdd:cert-manager:feature/azure-workload-identity?expand=1 ? I haven't tested it yet, just wrote it in accordance with your suggestion :)

Reply:
@weisdd  Yeah that looks to be about what I was thinking. Personally I would just pass it `opt.ClientID`/`managedIdentity.ClientID` _directly_ since the rest of the `adal.ManagedIdentityOptions` struct (ie `IdentityResourceID`) is useless to it.

Reply:
Some other thoughts:

I think(?; see further) it's fine to key off the prescense of the `AZURE_FEDERATED_TOKEN_FILE` variable. But at the same time, even if the token file is present we do allow the operator to specify a ServicePrincipal directly (ie. `azureDns.clientID` / the first ~12lines of  `getAuthorization`) and bypass AZWI. It's a little at odds with the below, but I think that's safe as the user would've explicitly needed to provide credentials. 

The `IdentityResourceId` however isn't used by AZWI, but it has a purpose in the context of pod-identity. If someone specified it, they'd likely(?) not be expecting AZWI. But because of the precense of the token variable we'd silently ignore the setting, even though we eagerly handle the SP case. 

That might be fine; I don't know why you'd be trying to use _both_ AZWI and podIdentity (outside of a cut-over maybe?). Perhaps an error should be raised if:

- `AZURE_FEDERATED_TOKEN_FILE` is set, and 
- [the actual token file is found(?), and] 
- `managedIdentity.resourceID` is set

Or AZWI should be _skipped_ if:

- `managedIdentity.resourceID` is set

I'd lean on the former (the error). Point being I think in the absence of a new _CRD or application level flag_ (to select MSI-"mode") the order of operations should be obvious/spelled out clearly. 

When deciding to use AZWI, is precense of the one variable enough? Perhaps we should check both _it_ and `AZURE_TENANT_ID` exist? Or perhaps that the file actually exists? If all conditions aren't met, _then_ fallback to the existing code? I can't see why someone would set the variable themselves so perhaps not a big deal...just thinking aloud.

Reply:
@pinkfloydx33 Sorry for the delay in my answers, several things overlapped at work, so quite busy.

Amongst the environment variables, I chose `AZURE_FEDERATED_TOKEN_FILE` just because it felt like the safest option. Potentially, cert-manager can have an option to override tenant-id through `tenantID`, so it's kind of fine to have other configuration means, whereas the federated token itself will always be read from disk as it's set to expire, simply incompatible with any static configuration option.

As for the sequence of authentication methods and additional errors, I'm not sure how tolerant cert-manager is expected to be for this type of misconfiguration (where you have certain settings set for a few different methods). It could feel like over-engineering, though if the code is expected to have a strict validation, it would be easy to add it.
For me, it's fine to just mention in the documentation that there's a certain set of supported authentication methods, and they're evaluated in a particular order. Plus, we already share some hints through log messages.

As for me passing `opt` instead of `opt.ClientID`, I just wanted to make it resemble the further call to `adal.NewServicePrincipalTokenFromManagedIdentity`, but it's fine to change that :)